File: src/vibeusage/providers/claude/status.py
"""Status page fetching utilities for providers."""
from __future__ import annotations

import json
from datetime import UTC
from datetime import datetime

from vibeusage.core.http import fetch_url
from vibeusage.models import ProviderStatus
from vibeusage.models import StatusLevel


async def fetch_statuspage_status(url: str) -> ProviderStatus:
    """Fetch status from a Statuspage.io status endpoint.

    Args:
        url: Statuspage.io API status URL

    Returns:
        ProviderStatus with current status
    """
    content = await fetch_url(url)
    if not content:
        return ProviderStatus.unknown()

    try:
        data = json.loads(content)
    except json.JSONDecodeError:
        return ProviderStatus.unknown()

    # Statuspage.io indicator mapping
    # "none" = operational, "minor" = degraded, "major" = partial_outage, "critical" = major_outage
    indicator = data.get("status", {}).get("indicator")

    level = _indicator_to_level(indicator)

    # Get description
    description = data.get("status", {}).get("description")

    return ProviderStatus(
        level=level,
        description=description,
        updated_at=datetime.now(UTC),
    )


def _indicator_to_level(indicator: str | None) -> StatusLevel:
    """Convert Statuspage.io indicator to StatusLevel."""
    mapping = {
        "none": StatusLevel.OPERATIONAL,
        "minor": StatusLevel.DEGRADED,
        "major": StatusLevel.PARTIAL_OUTAGE,
        "critical": StatusLevel.MAJOR_OUTAGE,
    }

    return mapping.get(indicator, StatusLevel.UNKNOWN)

File: vibeusage-go/internal/provider/status.go
package provider

import (
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/joshuadavidthomas/vibeusage/internal/models"
)

// FetchStatuspageStatus fetches status from a Statuspage.io endpoint.
func FetchStatuspageStatus(url string) models.ProviderStatus {
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return models.ProviderStatus{Level: models.StatusUnknown}
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return models.ProviderStatus{Level: models.StatusUnknown}
	}

	var data struct {
		Status struct {
			Indicator   string `json:"indicator"`
			Description string `json:"description"`
		} `json:"status"`
	}
	if err := json.Unmarshal(body, &data); err != nil {
		return models.ProviderStatus{Level: models.StatusUnknown}
	}

	level := indicatorToLevel(data.Status.Indicator)
	now := time.Now().UTC()
	return models.ProviderStatus{
		Level:       level,
		Description: data.Status.Description,
		UpdatedAt:   &now,
	}
}

func indicatorToLevel(indicator string) models.StatusLevel {
	switch strings.ToLower(indicator) {
	case "none":
		return models.StatusOperational
	case "minor":
		return models.StatusDegraded
	case "major":
		return models.StatusPartialOutage
	case "critical":
		return models.StatusMajorOutage
	default:
		return models.StatusUnknown
	}
}
